### DISCUSSION TITLE
Windows Captive AIR app crashing with exception 0xc0000005, any ideas?

### DISCUSSION THREAD

**Discussion by FliplineStudios** [airsdk/Adobe-Runtime-Support#3237] [Q&A]
One of our users has suddenly started getting crashes at startup for our PC game.  They're using Windows 11, and they were previously able to play the game without any problems.  This month whenever they try to launch the game they will either briefly see a black screen at startup or just their mouse cursor showing a spinning icon, then an immediate crash.  In the Event Viewer they see this information:

```
Faulting application name: Papa's Freezeria Deluxe.exe, version: 0.0.0.0, time stamp: 0x634977e1
Faulting module name: Adobe AIR.dll, version: 33.1.1.935, time stamp: 0x63497cd2
Exception code: 0xc0000005
Fault offset: 0x005814fd
Faulting process id: 0x0x5D40
Faulting application start time: 0x0x1DAA2360CC2CC85
Faulting application path: C:\Program Files (x86)\Steam\steamapps\common\Papa's Freezeria Deluxe\Papa's Freezeria Deluxe.exe
Faulting module path: C:\Program Files (x86)\Steam\steamapps\common\Papa's Freezeria Deluxe\Adobe AIR\Versions\1.0\Adobe AIR.dll
Report Id: 6429ca53-b001-4f3a-b416-1fb6e73cb83b
Faulting package full name:
Faulting package-relative application ID:
```

Not sure if that "Fault offset" offers any clues as to what's happening here.  This game is using the 32-bit captive runtime of 33.1.1.935, so it's not one of the newer AIR releases.  We use GPU renderMode for vector graphics, not using Stage3D, do not use any video or cameras, and the game uses DirectX 9 (using -swf-version=37 to force this).  Can't think of anything peculiar we're doing during startup, we load a couple external files, start playing music from the library and attach some MovieClips with vector graphics.  We haven't updated the game for over a year, so nothing has changed on our end.

The strange part is that the game played perfectly fine for them for months, and it was just recently that it suddenly can't open for them anymore, even after uninstalling/reinstalling from Steam.  We also do not load their game progress data right at startup, so it wouldn't be an issue with something changing/corrupting there causing an issue.

We've gone through the usual suspects, and ruled out it being caused by Discord (since Discord streaming/capture still causes Windows AIR apps to crash at launch unless you mess with the default settings), they've verified all Steam files, tried installing to different drives, checked for Windows updates, checked for driver updates, no antivirus running, tried launching the exe directly from the folder instead of launching via Steam, did an SFC scan, no luck with anything.

Does anyone have any ideas what could suddenly be causing this for this user?  We can't duplicate it ourselves (the game is working fine on our Windows 10 and Windows 11 machines) which makes it tricky to try newer AIR versions to test, though I didn't notice anything in the release notes for newer versions that might refer to this issue.  We're a bit stumped on what's happening, especially on why it only just started recently...

Any thoughts or ideas for things our user could test would be appreciated!

**Comment by ajwfrost** [comment]
> Not sure if that "Fault offset" offers any clues as to what's happening here

It _looks_ like this offset is within a function that resamples audio, which could happen either when processing data from the microphone, or when formatting it for output to the Windows wave audio device. So it might be good to get information about their speaker set-up or see whether they see different behaviour if they change their sound devices, disable all output, etc..? Or are you able to send a test version perhaps that doesn't use the microphone, or that has sound playback disabled?

If they're able to capture a minidump file, we could see if we're able to spot the issue from that?

thanks

**Reply by FliplineStudios** [reply]
Thanks, it did end up being related to audio playback!  For some reason, playing audio through their normal PC speakers (connected via audio cable) would now cause the game to crash at startup.  When they switched to headphones or the speakers built into their monitor, the game was able to play fine and played audio.  

They mentioned other programs and games worked fine when using the PC speakers, it was just ours that would crash if their PC was using those speakers (though the game used to play fine with them previously).

**Reply by ajwfrost** [reply]
That's interesting! Do you think they might be open to receiving a custom DLL that would output information about this and to allow us to try to work out the actual root cause?
thanks

**Reply by FliplineStudios** [reply]
I just heard back and they'd be up for giving it a try.  Thanks!

**Reply by Fancy2209** [reply]
Could it be their audio device format? I saw a person that was having a issue in a different AIR game where AIR failed to start because their audio device format was set to `24 bit, 384000 Hz`, and by setting it to `16 bit, 48000 Hz`, it was fixed.

**Reply by FliplineStudios** [reply]
I just heard from a different user having a similar crash:

```
Faulting module name: Adobe AIR.dll, version: 33.1.1.935, time stamp: 0x63497cd2
Exception code: 0xc0000095
Fault offset: 0x0047459d
Faulting process id: 0x0x180C
```

They also reported one Fault Offset of `0x0047451e`, but have multiple crashes that are all `0x0047459d`

Any idea if that Fault Offset is also audio related?  They've mentioned they've tried switching between speaker options and headphones, but it hasn't solved it for them...

EDIT: Just noticed this is a different Exception code as well: 0xc0000095

**Comment by dominicseo-nsuslab** [comment]
Hello ajwfrost,

We have encountered an application crash when calling flash.media.Sound.play() on a loaded or embedded audio asset. Invoking sound.play() after loading an audio resource causes the application to terminate without any error message. It appears that the SDK’s audio initialization or device selection logic may not properly handle disabled or unavailable audio devices—some users report that disabling their audio device in Windows 11 prevents the crash and allows normal playback ([url](https://answers.microsoft.com/en-us/windows/forum/all/windows-11-disable-audio-device-like-in-windows-10/3ad8db21-0227-40e2-b2f0-6ee1ba09fb46))

To reproduce the issue:

Instantiate a flash.media.Sound object in an AS3 project.
Load or embed an audio asset.
Call sound.play() and observe the immediate crash.

Below is an excerpt from a user’s crash dump for your reference:
```

FILE_IN_CAB:  GGnet.exe.20360 (1).dmp

NTGLOBALFLAG:  0

APPLICATION_VERIFIER_FLAGS:  0

CONTEXT:  (.ecxr)
rax=0000000000000000 rbx=00000000000007ec rcx=0000000000000000
rdx=0000000000000000 rsi=0000000000000001 rdi=0000000000000000
rip=00007ffa5b2b44ee rsp=0000001c493cabe0 rbp=0000001c493cac48
 r8=0000000000000001  r9=00000000000007ec r10=00007ffae84cf518
r11=0000000000000000 r12=0000000000000002 r13=0000017a66464b04
r14=0000017a418bf898 r15=0000017a61556fe0
iopl=0         nv up ei pl nz na pe nc
cs=0033  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00010200
Adobe_AIR+0x7444ee:
00007ffa`5b2b44ee 66890451        mov     word ptr [rcx+rdx*2],ax ds:00000000`00000000=????
Resetting default scope

EXCEPTION_RECORD:  (.exr -1)
ExceptionAddress: 00007ffa5b2b44ee (Adobe_AIR+0x00000000007444ee)
   ExceptionCode: c0000005 (Access violation)
  ExceptionFlags: 00000000
NumberParameters: 2
   Parameter[0]: 0000000000000001
   Parameter[1]: 0000000000000000
Attempt to write to address 0000000000000000

PROCESS_NAME:  GGnet.exe

WRITE_ADDRESS:  0000000000000000 

ERROR_CODE: (NTSTATUS) 0xc0000005 - 0x%p               0x%p                        .          %s              .

EXCEPTION_CODE_STR:  c0000005

EXCEPTION_PARAMETER1:  0000000000000001

EXCEPTION_PARAMETER2:  0000000000000000

STACK_TEXT:  
0000001c`493cabe0 00000000`00000000     : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`000007ec : Adobe_AIR+0x7444ee


STACK_COMMAND: ~0s; .ecxr ; kb

SYMBOL_NAME:  Adobe_AIR+7444ee

MODULE_NAME: Adobe_AIR

IMAGE_NAME:  Adobe AIR.dll

FAILURE_BUCKET_ID:  INVALID_POINTER_WRITE_c0000005_Adobe_AIR.dll!Unknown

OS_VERSION:  10.0.26100.1

BUILDLAB_STR:  ge_release

OSPLATFORM_TYPE:  x64

OSNAME:  Windows 10

IMAGE_VERSION:  51.1.3.8

FAILURE_ID_HASH:  {ce505e5e-24d2-b2c1-acc7-7492d146a9cf}

Followup:     MachineOwner
---------
```

Could you please advise on any known issues or recommended workarounds? 

Thank you for your assistance.

Best regards,
Dominic.seo

**Reply by Fancy2209** [reply]
I saw a person that was having the same issue, iirc the issue was their audio device format was set to 24 bit, 384000 Hz, The fix was setting it to 16 bit, 48000 Hz. This has to be done from the settings on windows

**Reply by ajwfrost** [reply]
Thanks .. this crash does look like it's in a similar area - audio resampling - and with the above hint we might be able to reproduce the problem here...

**Comment by FliplineStudios** [comment]
I do get a lot of reports from Windows users that our games will crash if they're using anything above 192,000 Hz sample rate for audio.  This is usually from players using an Audio DAC or amp on their PC which is set to a higher rate.

Definitely a lot of caveats we have to mention to PC users to prevent our AIR apps from crashing or losing input!  Especially for PC gamers who are used to their setup just working for other games.  (High audio sample rates cause crashes, Discord capturing/streaming immediately crashes, Discord overlays block mouse input, Overwolf overlays block mouse input or add lag, etc.)
